### PR TITLE
BTL simulation: OOT effects

### PR DESCRIPTION
**Pull request by casarsa**
#### PR description:

This PR modifies the BTL digitization to include the effects on time resolution due to the photon tails of earlier OOT hits in the same crystal. It implements the model proposed by A. Ledovskoy ([slides](https://indico.cern.ch/event/782676/contributions/3289163/attachments/1783938/2903520/190123_mtd_ledovskoy.pdf)). (@pmeridian @parbol )

The size of MTDSimHitData had to be increased in order for the information of previous BXs 
for both bar sides to be available in the BTLElectronicSim. The premixing configuration has been updated accordingly. (@makortel )

#### PR validation:

The new code has been tested in CMSSW_11_0_X_2019-11-18-2300 with the TTbar WF 22234 (PU200):

**step 2:**

MemoryReport> Peak virtual size 6280.26 Mbytes
 Key events increasing vsize: 
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[1] run: 1 lumi: 1 event: 1  vsize = 4296.21 deltaVsize = 0 rss = 3077.83 delta = 0
[3] run: 1 lumi: 1 event: 3  vsize = 5128.21 deltaVsize = 832 rss = 3166.95 delta = 89.1211
[88] run: 1 lumi: 1 event: 88  vsize = 6280.26 deltaVsize = 640 rss = 3146.59 delta = -20.3594
[12] run: 1 lumi: 1 event: 12  vsize = 5640.25 deltaVsize = 512.008 rss = 3204.43 delta = 37.4727
[90] run: 1 lumi: 1 event: 90  vsize = 6280.26 deltaVsize = 0 rss = 3078.7 delta = -67.8984
[89] run: 1 lumi: 1 event: 89  vsize = 6280.26 deltaVsize = 0 rss = 3203.86 delta = 57.2656
[88] run: 1 lumi: 1 event: 88  vsize = 6280.26 deltaVsize = 640 rss = 3146.59 delta = -34.625
TimeReport> Time report complete in 10911.9 seconds
Time Summary: 
 - Min event:   80.4885
 - Max event:   128.658
 - Avg event:   107.631
 - Total loop:  10882.3
 - Total init:  29.3701
 - Total job:   10911.9
 - EventSetup Lock:   0.00013113
 - EventSetup Get:   83.4706
 Event Throughput: 0.00918923 ev/s
 CPU Summary: 
 - Total loop:  9314.68
 - Total init:  13.381
 - Total extra: 0
 - Total job:   9328.3

 
**step 3**

 MemoryReport> Peak virtual size 10161.6 Mbytes
 Key events increasing vsize: 
[8] run: 1 lumi: 1 event: 8  vsize = 9139.4 deltaVsize = 32.0586 rss = 6650.14 delta = 89.9492
[3] run: 1 lumi: 1 event: 3  vsize = 9107.34 deltaVsize = 900.121 rss = 6560.19 delta = 155.914
[42] run: 1 lumi: 1 event: 42  vsize = 10161.6 deltaVsize = 1022.13 rss = 6870.45 delta = 310.262
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[44] run: 1 lumi: 1 event: 44  vsize = 10161.6 deltaVsize = 0 rss = 6714.96 delta = -155.488
[43] run: 1 lumi: 1 event: 43  vsize = 10161.6 deltaVsize = 0 rss = 6831.04 delta = -39.4141
[42] run: 1 lumi: 1 event: 42  vsize = 10161.6 deltaVsize = 1022.13 rss = 6870.45 delta = 68.0742
TimeReport> Time report complete in 24287.1 seconds
 Time Summary: 
 - Min event:   145.066
 - Max event:   412.86
 - Avg event:   237.186
 - Total loop:  24138
 - Total init:  138.612
 - Total job:   24287.1
 - EventSetup Lock:   0.000231981
 - EventSetup Get:   95.3623
 Event Throughput: 0.00414284 ev/s
 CPU Summary: 
 - Total loop:  20984.7
 - Total init:  101.39
 - Total extra: 0
 - Total job:   21096.3

### CODE REVIEW

**Comment by kpedro88** [comment]
please test

**Comment by makortel** [comment]
> The premixing configuration has been updated accordingly.

The changes themselves look fine. Did you check the impact on the pileup library disk size?

File: SimFastTiming/FastTimingCommon/src/BTLBarDeviceSim.cc [hunk 1]
```diff
@@ -84,36 +84,43 @@ void BTLBarDeviceSim::getHitsResponse(const std::vector<std::tuple<int, uint32_t
     // --- Get the simHit energy and convert it from MeV to photo-electrons
     float Npe = 1000. * hit.energyLoss() * LightYield_ * LightCollEff_ * PDE_;
 
-    // --- Get the simHit time of arrival
-    float toa = std::get<2>(hitRef);
```
Comment (kpedro88) [inline SimFastTiming/FastTimingCommon/src/BTLBarDeviceSim.cc]: why is this removed and replaced with `std::get<2>(hitRef)` multiple times in the lines below?

File: SimFastTiming/FastTimingCommon/src/BTLBarDeviceSim.cc [hunk 2]
```diff
@@ -84,36 +84,43 @@ void BTLBarDeviceSim::getHitsResponse(const std::vector<std::tuple<int, uint32_t
...
       distL = 0.5 * topo.pitch().first + 0.1 * hit.localPosition().x();
     }
 
-    double tR = toa + LightCollSlopeR_ * distR;
-    double tL = toa + LightCollSlopeL_ * distL;
+    double tR = std::get<2>(hitRef) + LightCollSlopeR_ * distR;
+    double tL = std::get<2>(hitRef) + LightCollSlopeL_ * distL;
+
+    // --- Accumulate in 15 buckets of 25ns (9 pre-samples, 1 in-time, 5 post-samples)
+    const int iBXR = std::floor(tR / bxTime_) + 9;
```
Comment (kpedro88) [inline SimFastTiming/FastTimingCommon/src/BTLBarDeviceSim.cc]: magic number 9 should be named constant

File: SimFastTiming/FastTimingCommon/src/BTLBarDeviceSim.cc [hunk 3]
```diff
@@ -84,36 +84,43 @@ void BTLBarDeviceSim::getHitsResponse(const std::vector<std::tuple<int, uint32_t
...
+    const int iBXL = std::floor(tL / bxTime_) + 9;
 
-    // --- Store the time of the first SimHit
-    if ((simHitIt->second).hit_info[1][0] == 0 || tR < (simHitIt->second).hit_info[1][0])
-      (simHitIt->second).hit_info[1][0] = tR;
+    // --- Right side
+    if (iBXR > 0 && iBXR < 15) {
```
Comment (kpedro88) [inline SimFastTiming/FastTimingCommon/src/BTLBarDeviceSim.cc]: magic number 15 should be named constant

File: SimFastTiming/FastTimingCommon/src/BTLElectronicsSim.cc [hunk 4]
```diff
@@ -77,6 +81,26 @@ void BTLElectronicsSim::run(const mtd::MTDSimHitDataAccumulator& input,
       float finalToA2 = finalToA1 + times[1];
       finalToA1 += times[0];
 
+      // --- Estimate the time uncertainty due to photons from earlier OOT hits in the current BTL cell
+      if (smearTimeForOOTtails_) {
+        float rate_oot = 0.;
+        // Loop on earlier OOT hits
+        for (unsigned int ibx = 0; ibx < 9; ++ibx) {
```
Comment (kpedro88) [inline SimFastTiming/FastTimingCommon/src/BTLElectronicsSim.cc]: magic number 9

File: SimFastTiming/FastTimingCommon/src/BTLElectronicsSim.cc [hunk 5]
```diff
@@ -77,6 +81,26 @@ void BTLElectronicsSim::run(const mtd::MTDSimHitDataAccumulator& input,
...
+          if ((it->second).hit_info[2 * iside][ibx] > 0.) {
+            float npe_oot = CLHEP::RandPoissonQ::shoot(hre, (it->second).hit_info[2 * iside][ibx]);
+            rate_oot += npe_oot * exp((it->second).hit_info[1 + 2 * iside][ibx] / ScintillatorDecayTime_) /
```
Comment (kpedro88) [inline SimFastTiming/FastTimingCommon/src/BTLElectronicsSim.cc]: can slightly speed up the math here by storing `ScintillatorDecayTimeInv_ = 1./ScintillatorDecayTime_` once and then multiplying by `ScintillatorDecayTimeInv_` (rather than dividing, which is slower)

File: SimFastTiming/FastTimingCommon/src/BTLTileDeviceSim.cc [hunk 6]
```diff
@@ -84,14 +84,16 @@ void BTLTileDeviceSim::getHitsResponse(const std::vector<std::tuple<int, uint32_
     if (smearLightCollTime_ > 0.)
       toa += CLHEP::RandGaussQ::shoot(hre, 0., smearLightCollTime_);
 
-    if (toa > bxTime_ || toa < 0)  //just consider BX==0
+    // Accumulate in 15 buckets of 25ns (9 pre-samples, 1 in-time, 5 post-samples)
+    const int iBX = std::floor(toa / bxTime_) + 9;
```
Comment (kpedro88) [inline SimFastTiming/FastTimingCommon/src/BTLTileDeviceSim.cc]: magic number 9

File: SimFastTiming/FastTimingCommon/src/BTLTileDeviceSim.cc [hunk 7]
```diff
@@ -84,14 +84,16 @@ void BTLTileDeviceSim::getHitsResponse(const std::vector<std::tuple<int, uint32_
...
+    if (iBX < 0 || iBX > 14)
```
Comment (kpedro88) [inline SimFastTiming/FastTimingCommon/src/BTLTileDeviceSim.cc]: magic number 14 = 15 - 1?

**Comment by fabiocos** [comment]
@casarsa please provide documentation about 

- the effect of the updated treatment of the OOT PU

- the backward compatibility of the results in case this is switched off

File: SimFastTiming/FastTimingCommon/src/BTLBarDeviceSim.cc (hunk 2 above)
Comment (casarsa) [reply SimFastTiming/FastTimingCommon/src/BTLBarDeviceSim.cc]: I've introduced a named constant

File: SimFastTiming/FastTimingCommon/src/BTLBarDeviceSim.cc (hunk 3 above)
Comment (casarsa) [reply SimFastTiming/FastTimingCommon/src/BTLBarDeviceSim.cc]: I've replaced 15 with a named constant

File: SimFastTiming/FastTimingCommon/src/BTLElectronicsSim.cc (hunk 4 above)
Comment (casarsa) [reply SimFastTiming/FastTimingCommon/src/BTLElectronicsSim.cc]: 9 has been replaced with a named constant

File: SimFastTiming/FastTimingCommon/src/BTLElectronicsSim.cc (hunk 5 above)
Comment (casarsa) [reply SimFastTiming/FastTimingCommon/src/BTLElectronicsSim.cc]: Done

File: SimFastTiming/FastTimingCommon/src/BTLTileDeviceSim.cc (hunk 6 above)
Comment (casarsa) [reply SimFastTiming/FastTimingCommon/src/BTLTileDeviceSim.cc]: Introduced named constant.

File: SimFastTiming/FastTimingCommon/src/BTLTileDeviceSim.cc (hunk 7 above)
Comment (casarsa) [reply SimFastTiming/FastTimingCommon/src/BTLTileDeviceSim.cc]: Rewritten with a named constant

File: SimFastTiming/FastTimingCommon/src/BTLBarDeviceSim.cc (hunk 1 above)
Comment (casarsa) [reply SimFastTiming/FastTimingCommon/src/BTLBarDeviceSim.cc]: I had a second thought on introducing a variable just to store a value which was already there

**Comment by casarsa** [comment]
> > The premixing configuration has been updated accordingly.
> 
> The changes themselves look fine. Did you check the impact on the pileup library disk size?

**Comment by casarsa** [comment]
My mistake

**Comment by casarsa** [comment]
> @casarsa please provide documentation about
> 
> * the effect of the updated treatment of the OOT PU

@pmeridian @parbol @fabiocos 
These are the time resolutions w/o and  w/ the OOT photon tails for BTL RECO hits with E>1 MeV in a sample of SingleMu PU200:
 
tres_pu200
tres_pu200_OOTtails


> * the backward compatibility of the results in case this is switched off

The new feature is controlled by a switch in the config file, the default is ON.
Here I compare the time resolution in a sample of TTbar events PU200 obtained in the current code and the new code w/ the OOT effect OFF:

cfr_OLD-NEW

**Comment by fabiocos** [comment]
please test workflow 20634.21

**Comment by makortel** [comment]
> From 100 events of WF 22261.97 with PU 200 I get:
> 
> Branch Name | Average Uncompressed Size (Bytes/Event) | Average Compressed Size (Bytes/Event)
> PMTDSimAccumulator_mix_FTLBarrel_DIGI. 2.96946e+06 2.39309e+06
> PMTDSimAccumulator_mix_FTLEndcap_DIGI. 1.33048e+06 775614

Thanks, so in total 3 MB/event (compressed). Could you also check the sizes without this PR? My previous estimate from 10_4_0_mtd5 is 1.1 MB/event, but that could be already out of date.

**Comment by casarsa** [comment]
> > From 100 events of WF 22261.97 with PU 200 I get:
> > Branch Name | Average Uncompressed Size (Bytes/Event) | Average Compressed Size (Bytes/Event)
> > PMTDSimAccumulator_mix_FTLBarrel_DIGI. 2.96946e+06 2.39309e+06
> > PMTDSimAccumulator_mix_FTLEndcap_DIGI. 1.33048e+06 775614
> 
> Thanks, so in total 3 MB/event (compressed). Could you also check the sizes without this PR? My previous estimate from 10_4_0_mtd5 is 1.1 MB/event, but that could be already out of date.

@makortel The estimate on 100 TTbar PU200 events with 11_0_X_2019-11-18-2300 w/o this PR is consistent with what you got:
PMTDSimAccumulator_mix_FTLEndcap_DIGI. 1.33048e+06 796149
PMTDSimAccumulator_mix_FTLBarrel_DIGI. 640898 393317

**Comment by fabiocos** [comment]
@pmeridian @parbol comments about this update? If we want it for the HLT TDR it should enter now.
In case there is the possibility to switch it off keeping the updated code.

**Comment by kpedro88** [comment]
Is it possible to tune the size of the premixed library? Without MTD and without truth information, the size was previously estimated at 12.7 MB/evt ([source](https://indico.cern.ch/event/711343/contributions/2964389/)). This change from 1 MB to 3 MB for MTD amounts to a ~15% increase in the overall library size, which is substantial.

**Comment by makortel** [comment]
> > > From 100 events of WF 22261.97 with PU 200 I get:
> > > Branch Name | Average Uncompressed Size (Bytes/Event) | Average Compressed Size (Bytes/Event)
> > > PMTDSimAccumulator_mix_FTLBarrel_DIGI. 2.96946e+06 2.39309e+06
> > > PMTDSimAccumulator_mix_FTLEndcap_DIGI. 1.33048e+06 775614
> > 
> > 
> > Thanks, so in total 3 MB/event (compressed). Could you also check the sizes without this PR? My previous estimate from 10_4_0_mtd5 is 1.1 MB/event, but that could be already out of date.
> 
> The estimate on 100 TTbar PU200 events with 11_0_X_2019-11-18-2300 w/o this PR is consistent with what you got:
> PMTDSimAccumulator_mix_FTLEndcap_DIGI. 1.33048e+06 796149
> PMTDSimAccumulator_mix_FTLBarrel_DIGI. 640898 393317

@casarsa Thanks! Increasing the compressed event size on disk from 1.1 MB to 3 MB is probably fine now. The total now is 45 MB, dominated by MC truth whose inclusion should become unnecessary. At that point the 3 MB/event could become something to be optimized (to compare, in 10_4_0_mtd5 HGCal consumed 8.5 MB/event (to be optimized as well), and tracker 2.8 MB/event. Run2 premixing at PU50 is 3.2 MB/event).

Ok, I see @kpedro88 already asked about reducing the size.

**Comment by kpedro88** [comment]
+upgrade
size reduction in the premix library should be attempted in a followup PR

**Comment by fabiocos** [comment]
@civanch any further comment?

**Comment by civanch** [comment]
+1

**Comment by fabiocos** [comment]
+1

**Comment by fabiocos** [comment]
@casarsa @kpedro88 @makortel strictly speaking this PR causes the premix library to be backward incompatible as far as MTD is concerned. I doubt this is a real practical issue, especially when moving from 10_6_X to 11_0_X, but it is important to have it clear

**Comment by kpedro88** [comment]
@fabiocos indeed, that should be fine because phase 2 premix libraries have not been centrally generated yet

**Comment by fabiocos** [comment]
@kpedro88 moving to a completely new release should be fine as well in my view, at the end of the day it is not a final data format but an intermediate technical one, to be discussed at least. Anyway I am afraid we have an issue with this PR and premix, as it is again crashing in the IB, I am making a cross check.

**Comment by fabiocos** [comment]
@casarsa @kpedro88 This is causing a crash in premixing workflow 20634.99 whose solution does not look entirely trivial. At this point, in order not to hold anymore the pre13 validation, I revert this PR to be possibly considered in a fix round after further careful validation.
